### PR TITLE
Certificate Accessor implementation

### PR DESCRIPTION
**Pull request by stark-dev**
This PR is still WIP. It includes code from @jana-rapava and @stark-dev on both the accessor and the library.
Unit tests are still missing, just submitting for a first code review.

TODO:
- unit test implementation
- ~~move string commands common to both the accessor and the library to include/fty-certificate-generator.h, as @jimklimov suggested~~

### CODE REVIEW

File: include/fty-certificate-generator.h [hunk 1]
```diff
@@ -25,6 +25,9 @@
 //  Include the project library file
 #include "fty_certificate_generator_library.h"
 
-//  Add your own public definitions here, if you need them
+// common command list
+static constexpr const char* GENERATE_SELFSIGNED_CERTIFICATE = "GENERATE_SELFSIGNED_CERTIFICATE";
```
Comment (perrettecl) [inline include/fty-certificate-generator.h]: need to be protected inside the namespace

File: src/certgen_accessor.cc [hunk 2]
```diff
@@ -19,16 +19,91 @@
...
+        const std::string & serviceName
+    ) const
+    {
+        fty::Payload payload;
+        try
+        {
+            payload = sendCommand(GENERATE_SELFSIGNED_CERTIFICATE, {serviceName});
+        }
+        catch (const std::exception& e)
+        {
```
Comment (perrettecl) [inline src/certgen_accessor.cc]: Do not log anything at library level. The user of the library should decide what to do with the exception

File: src/certgen_accessor.cc [hunk 3]
```diff
@@ -19,16 +19,91 @@
...
+    std::string CertGenAccessor::generateCsr(const std::string & serviceName) const
+    {
+        fty::Payload payload;
+        try
+        {
+            payload = sendCommand(GENERATE_CSR, {serviceName});
+        }
+        catch (const std::exception& e)
+        {
+            log_info("certificate generator accessor returned '%s'", e.what());
```
Comment (perrettecl) [inline src/certgen_accessor.cc]: Do not log anything at library level. The user of the library should decide what to do with the exception

File: src/certgen_accessor.cc [hunk 4]
```diff
@@ -19,16 +19,91 @@
...
+    ) const
+    {
+        fty::Payload payload;
+        try
+        {
+            payload = sendCommand(IMPORT_CERTIFICATE, {serviceName, cert});
+        }
+        catch (const std::exception& e)
+        {
+            log_info("certificate generator accessor returned '%s'", e.what());
```
Comment (perrettecl) [inline src/certgen_accessor.cc]: Do not log anything at library level. The user of the library should decide what to do with the exception

File: src/certgen_accessor.cc [hunk 5]
```diff
@@ -19,16 +19,91 @@
...
+        if(recMessage[0] == "ERROR")
+        {
+            // error - throw exception
+            if(recMessage.size() == 2)
+            {
+                throw std::runtime_error(recMessage.at(1));
+            }
+            else
+            {
+                throw std::runtime_error("Missing data for error");
```
Comment (perrettecl) [inline src/certgen_accessor.cc]: "Unknown error"

File: include/certgen_accessor.h [hunk 6]
```diff
@@ -16,15 +17,78 @@
...
     with this program; if not, write to the Free Software Foundation, Inc.,
     51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
-    =========================================================================
+    ========================================================================
 */
 
 #ifndef CERTGEN_ACCESSOR_H_INCLUDED
 #define CERTGEN_ACCESSOR_H_INCLUDED
 
+#include <fty_log.h>
```
Comment (perrettecl) [inline include/certgen_accessor.h]: Log is not needed in the header (not even in cpp as we do not log anything at library level)

File: include/certgen_accessor.h [hunk 7]
```diff
@@ -16,15 +17,78 @@
...
+#include <fty_common_mlm.h>
+
+#include <vector>
+#include <cxxtools/serializationinfo.h>
```
Comment (perrettecl) [inline include/certgen_accessor.h]: Should be in cpp as it's not part of any public definition.

Comment (perrettecl) [reply include/certgen_accessor.h]: Even in cpp we do not use it so far

File: src/certgen_certificate_generator_server.cc [hunk 8]
```diff
@@ -247,6 +249,31 @@ namespace certgen
         return config;
     }
 
+    // read config helper function
+    static CertificateGeneratorConfig generateConfig(const std::string & serviceName)
```
Comment (perrettecl) [inline src/certgen_certificate_generator_server.cc]: Maybe batter name "readConfig" or "getConfig"

File: src/certgen_certificate_generator_server.cc [hunk 9]
```diff
@@ -280,41 +307,83 @@ namespace certgen
...
+
+        std::string serviceName (params[0]);
+        CertificateGeneratorConfig certgenConfig = generateConfig(serviceName);        
+
+        fty::CertificateConfig config = loadConfig (certgenConfig.version(), certgenConfig.certConf());
+
+        std::string csrPem;
+
+        // TODO should we implement copy ctor on Keys??
+        m_csrKey.reset(new Keys(generateKeys(certgenConfig.keyConf())));
```
Comment (perrettecl) [inline src/certgen_certificate_generator_server.cc]: m_csrPendings => std::map<ServiceName, CsrX509>

For this you could compare and you do not need to use ptr

File: src/certgen_certificate_generator_server.cc [hunk 10]
```diff
@@ -280,41 +307,83 @@ namespace certgen
...
+        std::string certPem (params[1]);
+
+        CertificateGeneratorConfig certgenConfig = generateConfig(serviceName);        
+
+        fty::CertificateConfig config = loadConfig (certgenConfig.version(), certgenConfig.certConf());
+
+        CertificateX509 tmpCert(certPem);
+
+        // TODO check only public key??
+        if (tmpCert.getPublicKey() != m_csrKey->getPublicKey())
```
Comment (perrettecl) [inline src/certgen_certificate_generator_server.cc]: Yes

File: src/certgen_certificate_generator_server.h [hunk 11]
```diff
@@ -48,11 +49,8 @@ namespace certgen
             std::string handleGenerateCSR(const fty::Payload & params);
             std::string handleImportCertificate(const fty::Payload & params);
 
-        public:
-            //Command list
-            static constexpr const char* GENERATE_SELFSIGNED_CERTIFICATE = "GENERATE_SELFSIGNED_CERTIFICATE";
-            static constexpr const char* GENERATE_CSR = "GENERATE_CSR";
-            static constexpr const char* IMPORT_CERTIFICATE = "IMPORT_CERTIFICATE";
+            // attributes
+            std::unique_ptr<fty::Keys> m_csrKey; // needed to import an existing certificate correctly
```
Comment (perrettecl) [inline src/certgen_certificate_generator_server.h]: m_csrPendings => std::map<ServiceName, CsrX509>

**Review by perrettecl — CHANGES_REQUESTED** [review]

File: include/certgen_accessor.h [hunk 12]
```diff
@@ -16,15 +17,78 @@
...
+         */
+        void generateSelfCertificateReq(const std::string & serviceName) const;
+
+        /**
+         * @brief generate certificate signing request (CSR)
+         * @param serviceName name of the service
+         * 
+         * @return CSR PEM
+         */
+        std::string generateCsr(const std::string & serviceName) const;
```
Comment (perrettecl) [inline include/certgen_accessor.h]: Return CSR object

**Review by perrettecl — CHANGES_REQUESTED** [review]

File: src/certgen_certificate_generator_server.cc [hunk 13]
```diff
@@ -327,22 +325,19 @@ namespace certgen
...
-        m_csrKey.reset(new Keys(generateKeys(certgenConfig.keyConf())));
+        fty::CsrX509 csr = CsrX509::generateCsr(keyPair, config);
 
-        try
-        {
-            CsrX509 csr = CsrX509::generateCsr(*(m_csrKey.get()), config);
-            csrPem = csr.getPem();
-        }
-        catch(std::runtime_error &e)
+        // replace existing csr for a given service, if any
```
Comment (stark-dev) [inline src/certgen_certificate_generator_server.cc:332]: I could't find a better solution for this, there is a replace function for std::map [`insert_or_assign`](https://en.cppreference.com/w/cpp/container/map/insert_or_assign), but it's only available in std >= c++17 and gcc >= 5.x, I guess

Comment (perrettecl) [reply src/certgen_certificate_generator_server.cc:332]: Good to me.

**Review by perrettecl — APPROVED** [review]

**Pull request merged**